File: src/main/java/me/ag2s/epublib/domain/TitledResourceReference.java
package me.ag2s.epublib.domain;

import java.io.Serializable;
import me.ag2s.epublib.util.StringUtil;

public class TitledResourceReference extends ResourceReference implements Serializable {
   private static final long serialVersionUID = 3918155020095190080L;
   private String fragmentId;
   private String title;

   /** @deprecated */
   @Deprecated
   public TitledResourceReference(Resource resource) {
      this(resource, (String)null);
   }

   public TitledResourceReference(Resource resource, String title) {
      this(resource, title, (String)null);
   }

   public TitledResourceReference(Resource resource, String title, String fragmentId) {
      super(resource);
      this.title = title;
      this.fragmentId = fragmentId;
   }

   public String getFragmentId() {
      return this.fragmentId;
   }

   public void setFragmentId(String fragmentId) {
      this.fragmentId = fragmentId;
   }

   public String getTitle() {
      return this.title;
   }

   public void setTitle(String title) {
      this.title = title;
   }

   public String getCompleteHref() {
      return StringUtil.isBlank(this.fragmentId) ? this.resource.getHref() : this.resource.getHref() + '#' + this.fragmentId;
   }

   public Resource getResource() {
      if (this.resource != null && this.title != null) {
         this.resource.setTitle(this.title);
      }

      return this.resource;
   }

   public void setResource(Resource resource, String fragmentId) {
      super.setResource(resource);
      this.fragmentId = fragmentId;
   }

   public void setResource(Resource resource) {
      this.setResource(resource, (String)null);
   }
}

File: src/main/kotlin/me/ag2s/epublib/domain/TitledResourceReference.kt
package me.ag2s.epublib.domain

import java.io.Serializable
import me.ag2s.epublib.util.StringUtil

open class TitledResourceReference(resource: Resource?, private var title: String?, private var fragmentId: String? = null) : ResourceReference(resource), Serializable {
    @Deprecated("Deprecated in original API") constructor(resource: Resource?) : this(resource, null)
    fun getFragmentId(): String? = fragmentId
    fun setFragmentId(fragmentId: String?) { this.fragmentId = fragmentId }
    fun getTitle(): String? = title
    fun setTitle(title: String?) { this.title = title }
    fun getCompleteHref(): String? = if (StringUtil.isBlank(fragmentId)) resource!!.getHref() else resource!!.getHref() + '#' + fragmentId
    override fun getResource(): Resource? { if (resource != null && title != null) resource!!.setTitle(title); return resource }
    fun setResource(resource: Resource?, fragmentId: String?) { super.setResource(resource); this.fragmentId = fragmentId }
    override fun setResource(resource: Resource?) { setResource(resource, null) }
    companion object { private const val serialVersionUID = 3918155020095190080L }
}
